libglusterfs/src/glusterfs/gf-buf.h

```diff
@@ -0,0 +1,160 @@
+#ifndef _GF_BUF_H_
+#define _GF_BUF_H_
+
+#include "glusterfs/common-utils.h"
+
+#ifndef GF_BUF_STACK_THRESHOLD
+#define GF_BUF_STACK_THRESHOLD 64
+#endif
+
+typedef struct gf_buf {
+    unsigned long heap : 1;
+    unsigned long allocated : __BITS_PER_LONG - 1;
+    unsigned long used;
+    char *data;
+} gf_buf_t;
+
+void
+gf_buf_cleanup(struct gf_buf *);
+
+#define gf_buf_decl_null(var)                                                  \
+    gf_buf_t var __attribute__((__cleanup__(gf_buf_cleanup))) = {0, 0, 0, NULL}
+
+#define gf_buf_decl_empty(var)                                                 \
+    gf_buf_t var __attribute__((__cleanup__(gf_buf_cleanup))) = {0, 0, 1, ""}
+
+#define gf_buf_decl_string(var, str)                                           \
+    gf_buf_t var __attribute__((__cleanup__(gf_buf_cleanup))) = {              \
+        0, 0, strlen(str) + 1, (str)}
+
+#define __gf_buf_malloc(size) GF_MALLOC((size), gf_common_mt_char)
```

Comment:
I would use `static inline` definitions were possible. With functions it's easier to enforce types and detect errors. It also prevents variable names collisions.